### PR TITLE
HIVE-23363: Upgrade DataNucleus dependency to 5.2

### PR DESCRIPTION
**Pull request by belugabehr**

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/resources/package.jdo [hunk 1]
```diff
@@ -345,20 +345,20 @@
       <datastore-identity>
         <column name="CD_ID"/>
       </datastore-identity>
-      <field name="cols" table="COLUMNS_V2" >
```
Comment (ashutoshc) [inline standalone-metastore/metastore-server/src/main/resources/package.jdo:348]: Can you describe the need for this change?

File: standalone-metastore/metastore-server/src/main/resources/package.jdo [hunk 2]
```diff
@@ -345,20 +345,20 @@
...
+      <field name="cols" persistence-modifier="persistent" table="COLUMNS_V2">
         <collection element-type="MFieldSchema"/>
-        <foreign-key name="COLUMNS_V2_FK" delete-action="cascade"/>
```
Comment (ashutoshc) [inline standalone-metastore/metastore-server/src/main/resources/package.jdo:350]: We do want cascade-delete here. Any reason to remove it?

File: standalone-metastore/metastore-server/src/main/resources/package.jdo [hunk 3]
```diff
@@ -345,20 +345,20 @@
...
         <join>
-          <primary-key name="COLUMNS_PK">
```
Comment (ashutoshc) [inline standalone-metastore/metastore-server/src/main/resources/package.jdo:352]: Any reason to change the name here?

File: standalone-metastore/metastore-server/src/main/resources/package.jdo [hunk 4]
```diff
@@ -345,20 +345,20 @@
...
+          <primary-key name="SQL110922153006740">
+            <column name="CD_ID"/>
```
Comment (ashutoshc) [inline standalone-metastore/metastore-server/src/main/resources/package.jdo:352]: This probably is fine to do. Though, was it necessary?

**Comment by belugabehr** [comment]
@ashutoshc Let me see if I can address all of your questions with some background and context.  It took me a long time to get these changes to pass the unit tests.

So, these mappings, in some respect, don't really matter.  When HMS is started, users use the `schema-tool` to create the HMS schema for real.  Some of these mappings in the `jdo` file (like indexes) are only applied when unit testing because the unit tests build the schema via DN and `datanucleus.schema.autoCreateAll`.  For unit testing, the database backend is Apache Derby.  I changed the name of the index to match the Derby schema more closely.  In trying to debug these various errors, I was very confused at first about it complaining about "COLUMNS_PK".

https://github.com/apache/hive/blob/4942a7c0b4be3a5b0c889a89b903e9a70c57d494/standalone-metastore/metastore-server/src/main/sql/derby/hive-schema-4.0.0.derby.sql#L364

With that said, when I upgraded to DN 5.x, the unit tests would not pass.  I narrowed the issue down to this one table definition.  I tried several iterations to get success, but this is the one that worked.  I derived this solution by closely examining the docs on this topic.  It has an example that very closely aligns with this use case:

http://www.datanucleus.org/products/accessplatform/jpa/mapping.html#embedded_collection

It is a bit of a wonder looking at the existing JDO definition how this ever worked.

```
          <primary-key name="COLUMNS_PK">
            <column name="COLUMN_NAME"/>
          </primary-key>
```

This is not correct, this should be a compound primary key of CD_ID *and* COLUMN_NAME.  This is enforced by `SQL110922153006740` in the full schema.  As things currently stand, the COLUMN_NAME definition in the `jdo` file says that the COLUMN_NAME is not defined to be non-null.  This caused an error with Derby as it didn't allow creating a PRIMARY KEY on a field that could be null.

So, putting it all together, I came to the current solution.

File: standalone-metastore/metastore-server/src/main/resources/package.jdo (hunk 2 above)
Comment (belugabehr) [reply standalone-metastore/metastore-server/src/main/resources/package.jdo:350]:
```
If a foreign-key is specified (in MetaData) for the relation field then leave any deletion to the datastore to perform
``` 
I don't see any such relationship cascading relationship defined in the schema for Derby or MySQL, so DN should be doing it.

File: standalone-metastore/metastore-server/src/main/resources/package.jdo (hunk 1 above)
Comment (belugabehr) [reply standalone-metastore/metastore-server/src/main/resources/package.jdo:348]: Changed to:
```
<field name="cols" persistence-modifier="persistent" table="COLUMNS_V2">
```

Comment (belugabehr) [reply standalone-metastore/metastore-server/src/main/resources/package.jdo:348]: Just following the directions here:

http://www.datanucleus.org/products/accessplatform/jdo/mapping.html#embedded_collection

**Comment by belugabehr** [comment]
```
ALTER TABLE "APP"."COLUMNS_V2" ADD CONSTRAINT "COLUMNS_V2_FK1" FOREIGN KEY ("CD_ID") REFERENCES "APP"."CDS" ("CD_ID") ON DELETE NO ACTION ON UPDATE NO ACTION;
```

**Comment by belugabehr** [comment]
```
Foreign Keys

So we now have given the datastore control over the cascade deletion strategy for objects stored in these tables. Please be aware that JDO provides Dependent Fields as a way of allowing cascade deletion. The difference here is that Dependent Fields is controlled by DataNucleus, whereas foreign key delete actions are controlled by the datastore (assuming the datastore supports it even)
```

http://www.datanucleus.org/products/accessplatform/jdo/mapping.html#fk

**Comment by belugabehr** [comment]
Complete and merged through manual patch process 

https://issues.apache.org/jira/browse/HIVE-23363

**Pull request closed without merging**